docs/proposals/auth.md

```diff
@@ -0,0 +1,128 @@
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
+  pods to query for metrics within their own namespace.
+
+- separating querying from pushing: we should allow certain users to push, and
+  certain users to query, but not necessarily tie the two together
+
+- allowing different push scope permissions: we should be able to specify that
+  some users are only allowed to push metrics for the pods within their own
+  namespace (or given sufficient information in the future, possibly their own
+  RC), while others should be allowed to push for any object in the cluster.
+
+Proposed Solution
+-----------------
+
+Heapster will adopt an authentication and authorization model based around
+querying the Kubernetes model.  Authentication will support either client
+certificate and tokens using the Kubernetes API server tokenreview
+endpoint. Authorization will be done using the Kubernetes API server
+subjectaccessreviews endpoint.
+
+### Authentication ###
+
+Authentication will support both client certificates and tokens.  Client
+certificates will be supported by checking against a configurable CA (this
+will generally be the cluster CA that gets injected into the Heapster
+pod), while tokens will be checked against the Kubernetes API server using
+a `TokenReview`.
+
+The configuration will look as such:
+
+- `--authn-ca=/path/to/ca.crt`
+- `--auth-apiserver=https://$KUBE`
+
+### Authorization ###
+
+Authorization will work by matching Heapster concepts to Kubernetes RBAC
+concepts, and then performing `SubjectAccessReview` requests against the
+API server configured in `--auth-apiserver`.
+
+Concepts will map as such:
+
+#### Historical Queries #####
+
+**Verb**: get / **Subresource**: historical-metrics
+
+Namespace and Pod requests map to their respective resources.  Requests to
+get container metrics would map a request to get metrics on the
+corresponding pod.  Requests for node and system container metrics would
+simply map to cluster permissions to get historical-metrics on nodes.
+
+Getting by Pod UID would require permissions to get historical metrics on
+pods in any namespace, since we cannot tell namespace from the pod UID.
+
+Geting pods by a list would require making multiple requests, although for
+the sake of performance, it may be desirable to just merge them into
+`namespace=$NS, resource=pod, name=""` request.
+
+#### Model Queries ####
+
+**Verb**: get / **Subresource**: metrics
+
+These function more or less identically to historical queries, except that
+they don't have to worry about Pod UIDs.
+
+#### Push Creation ####
+
+**Verb**: create / **Subresource**: metrics, unprefixed-metrics
+
+The metrics subresource indicates that a user is allowed to push metrics
```

Comment:
Say I have an nginx pod, and it wants to push some metrics about itself, such as number of requests handled since starting.   Say that the pod has a credential with username  `$USERNAME`.

Is the recommended pattern:

1. the `nginx` pushes metrics prefixed with that username, like `$USERNAME-total-requests`.  Also, $USERNAME need to have a role binding which gives it `create metrics` permission. Or is it...
2.  the `nginx` pod  pushes unprefixed metrics, such as `total-requests`.  Also  $USERNAME needs to have a role binding which gives it `create unprefixed-metrics` permission.